src/mem/freelist.h

```diff
@@ -0,0 +1,297 @@
+    {
+      next_object = encode(key, next);
+      SNMALLOC_ASSERT(next == read_next(key));
+    }
+  };
+
+  /**
+   * Wrapper class that allows the keys for pointer encoding to be
+   * conditionally compiled.
+   */
+  class FreeObjectCursor
+  {
+    FreeObject* curr = nullptr;
+#ifdef CHECK_CLIENT
+    uintptr_t prev = 0;
+#endif
+
+    uintptr_t get_prev()
+    {
+#ifdef CHECK_CLIENT
+      return prev;
+#else
+      return 0;
+#endif
+    }
+
+    /**
+     * Updates the cursor to the new value,
+     * importantly this updates the key being used.
+     * Currently this is just the value of current before this call.
+     * Other schemes could be used.
+     */
+    void update_cursor(FreeObject* next)
+    {
+#ifdef CHECK_CLIENT
+#  ifndef NDEBUG
+      if (next != nullptr)
+      {
+        check_client(
+          !different_slab(prev, next),
+          "Heap corruption - free list corrupted!");
+      }
+#  endif
+      prev = address_cast(curr);
+#endif
+      curr = next;
+    }
+
+  public:
+    FreeObject* get_curr()
+    {
+      return curr;
+    }
+
+    /**
+     * Advance the cursor through the list
+     */
+    void move_next()
+    {
+#ifdef CHECK_CLIENT
+      check_client(
+        !different_slab(prev, curr), "Heap corruption - free list corrupted!");
+#endif
+      update_cursor(curr->read_next(get_prev()));
+    }
+
+    /**
+     * Update the next pointer at the location in the list pointed to
+     * by the cursor.
+     */
+    void set_next(FreeObject* next)
+    {
+      curr->store_next(next, get_prev());
+    }
+
+    /**
+     * Update the next pointer at the location in the list pointed to
+     * by the cursor, and move the cursor to that new value.
+     */
+    void set_next_and_move(FreeObject* next)
+    {
+      set_next(next);
+      update_cursor(next);
+    }
+
+    /**
+     * Resets the key to an initial value. So the cursor can be used
+     * on a new sequence.
+     */
+    void reset_cursor(FreeObject* next)
+    {
+#ifdef CHECK_CLIENT
+      prev = initial_key(next);
+#endif
+      curr = next;
+    }
+  };
+
+  /**
+   * Used to iterate a free list in object space.
+   *
+   * Checks signing of pointers
+   */
+  class FreeListIter
+  {
+  protected:
+    FreeObjectCursor front;
+
+  public:
+    /**
+     * Returns current value, without advancing the iterator.
+     */
+    void* peek()
+    {
+      return front.get_curr();
+    }
+
+    /**
+     * Checks if there are any more values to iterate.
+     */
+    bool empty()
+    {
+      return peek() == nullptr;
+    }
+
+    /**
+     * Moves the iterator on, and returns the current value.
+     */
+    void* take()
+    {
+      auto c = front.get_curr();
+      front.move_next();
+      return c;
+    }
+  };
+
+  /**
+   * Used to build a free list in object space.
+   *
+   * Checks signing of pointers
+   */
+  class FreeListBuilder : public FreeListIter
+  {
+    FreeObjectCursor end;
```

Comment:
If we just upcast, this won't go away even though (I suspect?) we never downcast from an `Iter` to a `Builder`.  Something something move constructor?

Reply:
It is so assignment from a builder to an iter "works". But I should perhaps codify that more carefully.

Reply:
Made inheritance private, and refactored slightly.

Reply:
Made inheritance non-public.